src/tracing.rs

```diff
@@ -68,5 +43,371 @@ pub fn setup(user_filters: &str) -> anyhow::Result<()> {
+    filter::Filtered<Option<fmt::Layer<Registry>>, filter::EnvFilter, Registry>,
+    Registry,
+>;
+
+impl LoggerHandle {
+    pub fn new(config: LoggerConfig, default: DefaultLoggerReloadHandle) -> anyhow::Result<Self> {
+        default.modify(|logger| update_default_logger(logger, &config.default))?;
+
+        let handle = Self {
+            config: Arc::new(RwLock::new(config)),
+            default,
+        };
+
+        Ok(handle)
+    }
+
+    pub async fn get_config(&self) -> LoggerConfig {
+        self.config.read().await.clone()
+    }
+
+    pub async fn update_config(&self, diff: LoggerConfigDiff) -> anyhow::Result<()> {
+        self.default
+            .modify(|logger| update_default_logger(logger, &diff.default))?;
+
+        self.config.write().await.update(diff);
+
+        Ok(())
+    }
+}
+
+#[derive(Clone, Debug, Default, Deserialize, Serialize)]
+pub struct LoggerConfig {
+    #[serde(flatten)]
+    default: DefaultLoggerConfig,
+}
+
+impl LoggerConfig {
+    pub fn with_top_level_directive(&mut self, log_level: Option<String>) -> &mut Self {
+        if self.default.log_level.is_some() && log_level.is_some() {
+            // TODO: Warn if both top-level `log_level` and `LoggerConfig::log_level` directives are used
+            eprintln!("TODO");
+        }
+
+        self.default.log_level = self.default.log_level.take().or(log_level);
+        self
+    }
+
+    pub fn update(&mut self, diff: LoggerConfigDiff) {
+        self.default.update(diff.default);
+    }
+}
+
+#[derive(Clone, Debug, Default, Deserialize, Serialize)]
+pub struct DefaultLoggerConfig {
+    log_level: Option<String>,
+    span_events: SpanEvents,
+    color: Color,
+}
+
+impl DefaultLoggerConfig {
+    pub fn update(&mut self, diff: DefaultLoggerConfigDiff) {
+        if let Some(log_level) = diff.log_level {
+            self.log_level = Some(log_level);
+        }
+
+        if let Some(span_events) = diff.span_events {
+            self.span_events = span_events;
+        }
+
+        if let Some(color) = diff.color {
+            self.color = color;
+        }
+    }
+}
+
+#[derive(Clone, Debug, Deserialize, Serialize, SmartDefault)]
+#[serde(from = "helpers::SpanEvents", into = "helpers::SpanEvents")]
+pub struct SpanEvents {
+    #[default(fmt::format::FmtSpan::NONE)]
+    events: fmt::format::FmtSpan,
+}
+
+impl From<fmt::format::FmtSpan> for SpanEvents {
+    fn from(events: fmt::format::FmtSpan) -> Self {
+        Self { events }
+    }
+}
+
+impl From<SpanEvents> for fmt::format::FmtSpan {
+    fn from(events: SpanEvents) -> Self {
+        events.events
+    }
+}
+
+#[derive(Copy, Clone, Debug, Deserialize, Serialize, SmartDefault)]
+#[serde(from = "helpers::Color", into = "helpers::Color")]
+pub enum Color {
+    #[default]
+    Auto,
+    Enable,
+    Disable,
+}
+
+impl Color {
+    pub fn to_bool(self) -> bool {
+        match self {
+            Self::Auto => io::stdout().is_terminal(),
```

Comment:
Ok, I think I figured how the check in `greeting.rs` works. Will fix.

Reply:
In `greeting.rs` we use functions from [`colored`](https://docs.rs/colored/latest/colored/) to actually color strings. It does the above checks internally. We'd have to do this explicitly here.

Reply:
Fixed this with the last update. It's _mostly_ the same as `greeting.rs`, except `Color::Enabled` and `Color::Disabled` unconditionally override any checks by `colored::control::ShouldColorize`.

I think, preferably, it should be the other way around: `ShouldColorize` should override config file, but `ShouldColorize` performs _a bunch_ of checks, and we either have to completely drop `color` field, or prioritize `color` over `ShouldColorize`. :/

> completely drop `color` field
I'll give it a second thought, though. Maybe it would be simpler.